Two new tutorials: materialized views + streaming ETL pipelines (#4921)
confluentinc/ksql: docs-md/app/etl.md

What is it?
----------

> A streaming ETL pipeline (sometimes called a “streaming data pipeline”) is a set of software services that ingests events, transforms them, and loads them into destination storage systems. It’s often the case that you have data in one place and want to move it to another as soon as you receive it, but you need to make some changes to the data as you transfer it. Maybe you need to do something simple, like transform the events to strip out any personally identifiable information. Sometimes you may need to do something more complex, like enrich the events by joining it with data from another system. Or perhaps you want to pre-aggregate the events to reduce how much data you send to the downstream systems. A streaming ETL pipeline lets you stream events between arbitrary sources and sinks helps you make changes to the data while it’s in-flight.

Reviewer: Some paragraph breaks might help readability. Also note typo in final sentence.
```suggestion
A streaming ETL pipeline (sometimes called a “streaming data pipeline”) is a set of software services that ingests events, transforms them, and loads them into destination storage systems. It’s often the case that you have data in one place and want to move it to another as soon as you receive it, but you need to make some changes to the data as you transfer it.

Maybe you need to do something simple, like transform the events to strip out any personally identifiable information. Sometimes you may need to do something more complex, like enrich the events by joining it with data from another system. Or perhaps you want to pre-aggregate the events to reduce how much data you send to the downstream systems.

A streaming ETL pipeline lets you stream events between arbitrary sources and sinks, and it helps you make changes to the data while it’s in-flight.
```

![hard](../img/etl-hard.png){: style="width:90%;"}

One way you might do this would be to capture the changelogs of upstream Postgres and MongoDB databases. The changelog can be stored in Kafka where a series of deployed programs transform, aggregate, and join the data together. The data can finally be streamed out to ElasticSearch for indexing. Many people build this sort of architecture, but could it be made simpler?

Why ksqlDB
----------

> Gluing all of the above services together is certainly a challenge. Along with your original databases and target analytical data store, you end up managing clusters for Kafka, connectors, your stream processors. It's hard to operate the entire stack as one. ksqlDB helps streamline how you write and deploy streaming data pipelines by boiling it down to just two things: storage (Kafka) and compute (ksqlDB).

Reviewer: ```suggestion
Gluing all of the above services together is certainly a challenge. Along with your original databases and target analytical data store, you end up managing clusters for Kafka, connectors, and your stream processors. It's challenging to operate the entire stack as one.

ksqlDB helps streamline how you write and deploy streaming data pipelines by boiling it down to just two things: storage (Kafka) and compute (ksqlDB).
```

      KSQL_KSQL_LOGGING_PROCESSING_TOPIC_AUTO_CREATE: "true"
      KSQL_CONNECT_GROUP_ID: "ksql-connect-cluster"
      KSQL_CONNECT_BOOTSTRAP_SERVERS: "broker:9092"
      KSQL_CONNECT_KEY_CONVERTER: "io.confluent.connect.avro.AvroConverter"
      KSQL_CONNECT_VALUE_CONVERTER: "io.confluent.connect.avro.AvroConverter"
      KSQL_CONNECT_KEY_CONVERTER_SCHEMA_REGISTRY_URL: "http://schema-registry:8081"
      KSQL_CONNECT_VALUE_CONVERTER_SCHEMA_REGISTRY_URL: "http://schema-registry:8081"
      KSQL_CONNECT_VALUE_CONVERTER_SCHEMAS_ENABLE: "false"
      KSQL_CONNECT_CONFIG_STORAGE_TOPIC: "ksql-connect-configs"
      KSQL_CONNECT_OFFSET_STORAGE_TOPIC: "ksql-connect-offsets"
      KSQL_CONNECT_STATUS_STORAGE_TOPIC: "ksql-connect-statuses"
      KSQL_CONNECT_CONFIG_STORAGE_REPLICATION_FACTOR: 1
      KSQL_CONNECT_OFFSET_STORAGE_REPLICATION_FACTOR: 1
      KSQL_CONNECT_STATUS_STORAGE_REPLICATION_FACTOR: 1
      KSQL_CONNECT_PLUGIN_PATH: "/usr/share/kafka/plugins"

  ksqldb-cli:
    image: confluentinc/ksqldb-cli:0.8.0
    container_name: ksqldb-cli
    depends_on:
      - broker
      - ksqldb-server
    entrypoint: /bin/sh
    tty: true
```

> There's a couple things to notice here. The Postgres image mounts the custom configuration file that we wrote. Postgres adds these configuration settings into its system-wide configuration. The environment variables we gave it also set up a blank database called `customers` along with a user named `postgres-user` that can access it.

Reviewer: ```suggestion
There are a couple things to notice here. The Postgres image mounts the custom configuration file that we wrote. Postgres adds these configuration settings into its system-wide configuration. The environment variables we gave it also set up a blank database called `customers` along with a user named `postgres-user` that can access it.
```

> We’ve also set up MongoDB as a replica set named `my-replica-set` Debezium requires that MongoDB run in this configuration to pick up changes from its oplog. In this case, we’re just running a single node replica set.

Reviewer: ```suggestion
We’ve also set up MongoDB as a replica set named `my-replica-set`. Debezium requires that MongoDB run in this configuration to pick up changes from its oplog. In this case, we’re just running a single node replica set.
```

Bring up the entire stack by running:

```
docker-compose up
```

### Create the customers table in Postgres

It's pretty common for companies to keep their customer data in a relational database. Let's model that information in a Postgres table. Start by logging into the container:

```
docker exec -it postgres /bin/bash
```

Log into Postgres as the user created by default:

```
psql -U postgres-user customers
```

Create a table that represents the customers. For simplicity's sake, we'll just model this with three columns: an id, a name, and the age of the person:

```sql
CREATE TABLE customers (id TEXT PRIMARY KEY, name TEXT, age INT);
```

Seed the table with some initial data:

```sql
INSERT INTO customers (id, name, age) VALUES ('5', 'fred', 34);
INSERT INTO customers (id, name, age) VALUES ('7', 'sue', 25);
INSERT INTO customers (id, name, age) VALUES ('2', 'bill', 51);
```

### Configure MongoDB for Debezium

Now that Postgres is setup, let's configure MongoDB. Start by logging into the container:

```
docker exec -it mongo /bin/bash
```

Log into the Mongo console using the username and password specified in the Docker Compose file:

```

> mongo -u mongo-user -p mongo-pw admin

Reviewer: consider using env vars passed to the container here, e.g. `MONGO_INITDB_ROOT_USERNAME`

```

Because MongoDB has been started as a replica set, it needs to be initiated. Run the following command to kick it off:

```javascript
rs.initiate()
```

Now that this node has become the primary in the replica set, we need to configure access so that Debezium can replicate changes remotely. Switch into the `config` database:

```
use config
```

Create a new role for Debezium. This role will enable the user that we create to access system-level collections, which are normally restricted:

```javascript
db.createRole({
    role: "dbz-role",
    privileges: [
        {
            resource: { db: "config", collection: "system.sessions" },
            actions: [ "find", "update", "insert", "remove" ]
        }
    ],
    roles: [
       { role: "dbOwner", db: "config" },
       { role: "dbAdmin", db: "config" },
       { role: "readWrite", db: "config" }
    ]
})
```

Switch into the `admin` database. We need to create our user here so that it can be authenticated:

```
use admin
```

> Create the user for Debezium. This user has `root` on the `admin` database, and can also access other databases needed for replication:

Reviewer: ```suggestion
Create the user for Debezium. This user has `root` on the `admin` database, and it can also access other databases needed for replication:
```

```

Do the same for shipments. Notice that the `order_id` references order ids we created in the previous collection.

```javascript
db.shipments.insert({"order_id": "17", "shipment_id": "75", "origin": "texas", "ts": "2020-04-04T19:20:00"})
db.shipments.insert({"order_id": "22", "shipment_id": "71", "origin": "iowa", "ts": "2020-04-04T12:25:00"})
db.shipments.insert({"order_id": "29", "shipment_id": "89", "origin": "california", "ts": "2020-04-05T13:21:00"})
db.shipments.insert({"order_id": "13", "shipment_id": "92", "origin": "maine", "ts": "2020-04-04T06:13:00"})
db.shipments.insert({"order_id": "15", "shipment_id": "95", "origin": "florida", "ts": "2020-04-04T01:13:00"})
```

### Start the Postgres and MongoDB Debezium source connectors

With all of our seed data in place, we can process it with ksqlDB. Connect to ksqlDB's server using its interactive CLI. Run the following from your host:

```
docker exec -it ksqldb-cli ksql http://ksqldb-server:8088
```

Before we issue more commands, instruct ksqlDB to start all queries from earliest point in each topic:

```
SET 'auto.offset.reset' = 'earliest';
```

> Now we can connect ask Debezium to stream Postgres' changelog into Kafka. Invoke the following command in ksqlDB. This creates a Debezium source connector and writes all of its changes to Kafka topics:

Reviewer: ```suggestion
Now we can ask Debezium to stream Postgres' changelog into Kafka. Invoke the following command in ksqlDB. This creates a Debezium source connector and writes all of its changes to Kafka topics:
```

```sql
CREATE SOURCE CONNECTOR customers_reader WITH (
    'connector.class' = 'io.debezium.connector.postgresql.PostgresConnector',
    'database.hostname' = 'postgres',
    'database.port' = '5432',
    'database.user' = 'postgres-user',
    'database.password' = 'postgres-pw',
    'database.dbname' = 'customers',
    'database.server.name' = 'customers',
    'database.history.kafka.bootstrap.servers' = 'broker:9092',
    'database.history.kafka.topic' = 'customers',
    'table.whitelist' = 'public.customers',
    'transforms' = 'unwrap',
    'transforms.unwrap.type' = 'io.debezium.transforms.ExtractNewRecordState',
    'transforms.unwrap.drop.tombstones' = 'false',
    'transforms.unwrap.delete.handling.mode' = 'rewrite'
);
```

> Notice that we specified an `unwrap` transform. By default, Debezium sends all events in an envelop that include many pieces of information about the change captured. Here, we only care about the value after it changed, so we instruct Kafka Connect to simply keep that information and discard the rest.

Reviewer: ```suggestion
Notice that we specified an `unwrap` transform. By default, Debezium sends all events in an envelope that includes many pieces of information about the change captured. Here, we only care about the value after it changed, so we instruct Kafka Connect to simply keep that information and discard the rest.
```

> Run another source connector to ingest the changes from MongoDB. We specify the same behavior for discarding the Debezium envelop:

Reviewer: ```suggestion
Run another source connector to ingest the changes from MongoDB. We specify the same behavior for discarding the Debezium envelope:
```

```sql
CREATE SOURCE CONNECTOR logistics_reader WITH (
    'connector.class' = 'io.debezium.connector.mongodb.MongoDbConnector',
    'mongodb.hosts' = 'mongo:27017',
    'mongodb.name' = 'my-replica-set',
    'mongodb.authsource' = 'admin',
    'mongodb.user' = 'dbz-user',
    'mongodb.password' = 'dbz-pw',
    'database.history.kafka.bootstrap.servers' = 'broker:9092',
    'database.history.kafka.topic' = 'logistics',
    'collection.whitelist' = 'logistics.*',
    'transforms' = 'unwrap',
    'transforms.unwrap.type' = 'io.debezium.connector.mongodb.transforms.ExtractNewDocumentState',
    'transforms.unwrap.drop.tombstones' = 'false',
    'transforms.unwrap.delete.handling.mode' = 'drop',
    'transforms.unwrap.operation.header' = 'true'
);
```

### Create the ksqlDB source streams

> For ksqlDB to be able to use the topics that Debezium created, we need to declare streams over it. Because we configured Kafka Connect with Schema Registry, we don't need to declare the schema of the data for the streams. It is simply inferred the schema that Debezium writes with.

Reviewer: ```suggestion
For ksqlDB to be able to use the topics that Debezium created, we need to declare streams over it. Because we configured Kafka Connect with Schema Registry, we don't need to declare the schema of the data for the streams. It simply infers the schema that Debezium writes with.
```
